Dockerfile

```diff
@@ -1,36 +1,30 @@
-FROM ubuntu:bionic
+FROM ubuntu:bionic-20191010
```

Comment:
Please note I changed to a specific version of bionic (latest one) because if they update bionic it will not build everything again.
If you keep pointing to `ubuntu:bionic` every time they update that tag it will run a full build of your Docker image again. It's also valid as they update bionic tag every few weeks and the advantage is that you have a more secure image. 
But this is to give you an example of optimising the build by binding the image to a specific tag.

Reply:
Sounds good not binding to a specific version. I just wanted to give an example of how to optimise. 
Cache save you lots of time, but for final release is better wait a bit longer for a full build for sure.
Shippable doesn't maintain cache, in fact. So to bypass you could push one base image to some Docker Registry and just download from there. It can save time as downloading is faster than building again.
What about all those packages being installed? You might need them for the other tools that use toolchain, right? What if you move the installation of those packages to that apps? Do you think it looses the idea of toolchain?
You need to see what's your real bottleneck? It's time spent in development cycle? Every cycle to test a change it takes centuries to build? So improve in that stage!
The point is you don't need to download all packages every time you want to test a small change, for sure.